File: SDEventsViewController.m
//
//  SDEventsViewController.m
//  ShralpTide2
//
//  Created by Michael Parlee on 9/23/13.
//
//

#import "SDEventsViewController.h"
#import "SDTideEvent.h"
#import "SDTideEventCell.h"
#import "SDTideFactory.h"

@interface SDEventsViewController ()

@end

@implementation SDEventsViewController

- (id)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil
{
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self) {
        // Custom initialization
    }
    return self;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    self.dateLabel.adjustsFontSizeToFitWidth = YES;
    self.dateLabel.translatesAutoresizingMaskIntoConstraints = NO;
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    
    [[NSNotificationCenter defaultCenter] removeObserver:self];
    NSDateFormatter *formatter = [[NSDateFormatter alloc] init];
    formatter.formatterBehavior = NSDateFormatterBehaviorDefault;
    formatter.dateStyle = NSDateFormatterFullStyle;
    self.dateLabel.text = [formatter stringFromDate:[self.tide startTime]];
    
    DLog(@"Scroll view frame width = %f", _chartScrollView.frame.size.width);
    _chartView.height = 40;
    _chartView.datasource = self;
    if ([_tide.startTime timeIntervalSince1970] > [[NSDate date] timeIntervalSince1970]) {
        _chartView.hoursToPlot = 24;
        _chartView.frame = CGRectMake(0, 0, 1200, _chartView.frame.size.height);
    } else {
        _chartView.hoursToPlot = 12;
        [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(redrawChart) name:kSDApplicationActivatedNotification object:nil];
    }
    
    DLog(@"Setting content width to %f",_chartView.frame.size.width);
    _chartScrollView.contentSize = _chartView.frame.size;
    [_chartScrollView addSubview:_chartView];
}

- (void)redrawChart
{
    [_chartView setNeedsDisplay];
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark Table View Delegate Methods

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath
{
    return tableView.bounds.size.height / 4;
}

#pragma mark Table View Data Source methods

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    NSDate *date = [self.tide startTime];
    NSCalendar *calendar = [NSCalendar autoupdatingCurrentCalendar];
    NSUInteger preservedComponents = (NSCalendarUnitYear | NSCalendarUnitMonth | NSCalendarUnitDay);
    date = [calendar dateFromComponents:[calendar components:preservedComponents fromDate:date]];
    date = [date dateByAddingTimeInterval:1440 * 60]; // add one day to get midnight tonight
    
    NSPredicate *predicate = [NSPredicate predicateWithFormat:@"eventTime < %@", date];
    NSArray *todaysEvents = [self.tide.events filteredArrayUsingPredicate:predicate];
    return [todaysEvents count];
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return 1;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    DLog(@"Cell for Row at Index Path %ld called by table %@", (long)indexPath.row, tableView);
    static NSString* reuseId = @"eventCell";
    SDTideEvent *event = (SDTideEvent*)self.tide.events[indexPath.row];
    SDTideEventCell *cell = [tableView dequeueReusableCellWithIdentifier:reuseId];
    cell.timeLabel.text = event.eventTimeNativeFormat;
    cell.heightLabel.text = [NSString stringWithFormat:@"%1.2f %@",event.eventHeight,self.tide.unitShort];
    cell.typeLabel.text = event.eventTypeDescription;
    return cell;
}

#pragma mark ChartViewDatasource Methods
-(SDTide *)tideDataToChart
{
    return _tide;
}

-(NSDate*)day
{
    NSDate *now = [NSDate date];
    if ([_tide.startTime timeIntervalSince1970] > [now timeIntervalSince1970]) {
        return _tide.startTime;
    } else {
        return now;
    }
}

-(int)page
{
    return 0;
}
@end

File: SDEventsViewController.swift
//
//  SDEventsViewController.swift
//  ShralpTide2
//
//  Created by Michael Parlee on 11/17/18.
//

import Foundation

@objc class SDEventsViewController: UIViewController {

    @IBOutlet fileprivate weak var dateLabel: UILabel! {
        didSet {
            self.dateLabel.adjustsFontSizeToFitWidth = true
            self.dateLabel.translatesAutoresizingMaskIntoConstraints = false
        }
    }
    @IBOutlet fileprivate var chartView: ChartView!
    @IBOutlet fileprivate weak var chartScrollView: UIScrollView!
    @IBOutlet fileprivate weak var bottomVerticalConstraint: NSLayoutConstraint!
    @objc var tide: SDTide!
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        NotificationCenter.default.removeObserver(self)
        let formatter = DateFormatter()
        formatter.formatterBehavior = DateFormatter.Behavior.default
        formatter.dateStyle = DateFormatter.Style.full
        self.dateLabel.text = formatter.string(from: self.tide.startTime)
        
        chartView.height = 40;
        chartView.datasource = self;
        chartView.hoursToPlot = 24;
        chartView.frame = CGRect(x: 0,
                                  y:0,
                                  width: chartScrollView.frame.size.width,
                                  height: chartView.frame.size.height
        )
        
        chartScrollView.contentSize = chartView.frame.size
        chartScrollView.addSubview(chartView)
    }
}

extension SDEventsViewController: ChartViewDatasource {
    var tideDataToChart: SDTide! {
        return self.tide
    }
    
    var day: Date! {
        return self.tide.startTime
    }
    
    var page: Int32 {
        return 0
    }
}

extension SDEventsViewController: UITableViewDelegate {
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return tableView.bounds.size.height / 4
    }
}

extension SDEventsViewController: UITableViewDataSource {
    static let reuseId = "eventCell"
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.tide.events(forDay: self.tide.startTime).count
    }
    
    func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let event = self.tide.events[indexPath.row]
        let cell = tableView.dequeueReusableCell(withIdentifier: SDEventsViewController.reuseId) as! SDTideEventCell
        cell.timeLabel.text = event.eventTimeNativeFormat
        cell.heightLabel.text = String(format: "%1.2f%@", event.eventHeight, self.tide.unitShort)
        cell.typeLabel.text = event.eventTypeDescription
        return cell
    }
}

